File: armorpaint/Sources/slot_brush.ts

class slot_brush_t {
	nodes: zui_nodes_t = zui_nodes_create();
	canvas: zui_node_canvas_t;
	image: image_t = null; // 200px
	image_icon: image_t = null; // 50px
	preview_ready: bool = false;
	id: i32 = 0;
}

let slot_brush_default_canvas: buffer_t = null;

function slot_brush_create(c: zui_node_canvas_t = null): slot_brush_t {
	let raw: slot_brush_t = new slot_brush_t();
	for (let brush of project_brushes) {
		if (brush.id >= raw.id) {
			raw.id = brush.id + 1;
		}
	}

	if (c == null) {
		if (slot_brush_default_canvas == null) { // Synchronous
			let b: buffer_t = data_get_blob("default_brush.arm")
			slot_brush_default_canvas = b;
		}
		raw.canvas = armpack_decode(slot_brush_default_canvas);
		raw.canvas.name = "Brush " + (raw.id + 1);
	}
	else {
		raw.canvas = c;
	}

	return raw;
}

File: paint/sources/slot_brush.c

#include "global.h"

buffer_t      *slot_brush_default_canvas = NULL;

slot_brush_t *slot_brush_create(ui_node_canvas_t *c) {
	slot_brush_t *raw  = GC_ALLOC_INIT(slot_brush_t, {0});
	raw->nodes         = ui_nodes_create();
	raw->preview_ready = false;
	raw->id            = 0;

	for (i32 i = 0; i < project_brushes->length; ++i) {
		slot_brush_t *brush = project_brushes->buffer[i];
		if (brush->id >= raw->id) {
			raw->id = brush->id + 1;
		}
	}

	if (c == NULL) {
		if (slot_brush_default_canvas == NULL) { // Synchronous
			buffer_t *b = data_get_blob("default_brush.arm");
			gc_unroot(slot_brush_default_canvas);
			slot_brush_default_canvas = b;
			gc_root(slot_brush_default_canvas);
		}
		raw->canvas = armpack_decode(slot_brush_default_canvas);
		raw->canvas = util_clone_canvas(raw->canvas); // Clone to create GC references

		i32 id            = (raw->id + 1);
		raw->canvas->name = string("Brush %d", id);
	}
	else {
		raw->canvas = util_clone_canvas(c);
	}

	return raw;
}
